dali/kernels/imgproc/color_manipulation/debayer/debayer_test.cc

```diff
@@ -0,0 +1,176 @@
+
+#include "dali/core/tensor_shape_print.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/dynamic_scratchpad.h"
+#include "dali/kernels/imgproc/color_manipulation/debayer/debayer.h"
+#include "dali/kernels/imgproc/color_manipulation/debayer/debayer_npp.h"
+#include "dali/kernels/scratch.h"
+#include "dali/pipeline/data/tensor_list.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/test/test_tensors.h"
+
+namespace dali {
+namespace kernels {
+namespace debayer {
+namespace test {
+
+constexpr cudaStream_t cuda_stream = 0;
+
+template <typename InOutT_, DALIDebayerAlgorithm alg_>
+struct DebayerTestParams {
+  using InOutT = InOutT_;
+  static constexpr DALIDebayerAlgorithm alg = alg_;
+};
+
+template <typename DebayerTestParamsT>
+class DebayerGpuTest : public ::testing::Test {
+ protected:
+  using InOutT = typename DebayerTestParamsT::InOutT;
+  using Kernel = NppDebayerKernel<InOutT>;
+  static constexpr int num_channels = 3;
+  static_assert(DebayerTestParamsT::alg == DALIDebayerAlgorithm::DALI_DEBAYER_BILINEAR_NPP);
+
+  void FillWithGradient(TensorListView<StorageCPU, InOutT, 3> rgb_batch) {
+    int max_val = std::numeric_limits<InOutT>::max();
+    static constexpr int num_channels = 3;
+    const auto &batch_shape = rgb_batch.shape;
+    for (int sample_idx = 0; sample_idx < rgb_batch.num_samples(); sample_idx++) {
+      int height = batch_shape[sample_idx][0];
+      int width = batch_shape[sample_idx][1];
+      auto rgb_sample = rgb_batch[sample_idx];
+      for (int h = 0; h < height; h++) {
+        for (int w = 0; w < width; w++) {
+          rgb_sample.data[h * num_channels * width + w * num_channels] = max_val * (w + 1) / width;
+        }
+      }
+      for (int h = 0; h < height; h++) {
+        for (int w = 0; w < width; w++) {
+          rgb_sample.data[h * num_channels * width + w * num_channels + 1] =
+              max_val * (h + 1) / height;
+        }
+      }
+      for (int h = 0; h < height; h++) {
+        for (int w = 0; w < width; w++) {
+          rgb_sample.data[h * num_channels * width + w * num_channels + 2] =
+              max_val * (width - w) / width;
+        }
+      }
+    }
+  }
+
+  void BayerSamples(TensorListView<StorageCPU, InOutT, 2> bayer_batch,
+                    TensorListView<StorageCPU, const InOutT, 3> rgb_batch) {
+    // Note that dali uses opncv's convention of naming the patterns, which
+    // looks at the 2x2 tile starting at the second row and column of the sensors matrix.
+    // We iterate over whole image, so the patterns are first transformed as if by looking
+    // at the tile starting at first column and row.
+    std::array<std::array<std::array<int, 2>, 2>, 4> pattern2channel{{
+        {{{0, 1}, {1, 2}}},  // bggr -> rggb -> 0112
+        {{{1, 0}, {2, 1}}},  // gbrg -> grbg -> 1021
+        {{{1, 2}, {0, 1}}},  // grbg -> gbrg -> 1201
+        {{{2, 1}, {1, 0}}}   // rggb -> bggr -> 2110
+    }};
```

Comment:
I think the braces are mismatched. Plain C array would be easier to read.

Reply:
```suggestion
    int pattern2channel[4][2][2] = {
        {{0, 1}, {1, 2}},  // bggr -> rggb -> 0112
        {{1, 0}, {2, 1}},  // gbrg -> grbg -> 1021
        {{1, 2}, {0, 1}},  // grbg -> gbrg -> 1201
        {{2, 1}, {1, 0}}   // rggb -> bggr -> 2110
    };
```

Reply:
done